### DISCUSSION TITLE
Open String Notes

### DISCUSSION THREAD

**Discussion by kyledecot** [CoderLine/alphaTab#917] [Ideas]
It would be very helpful if the AlphaTab API exposed either a getter or function to return/produce an array of `model.Note` objects that mapped to the open strings of a given track tuning w/ an optional duration. This would be useful in conjunction w/ `playNote`. 

As an alternative / more low level API  it would be useful to be able to create notes of arbitrary frequencies / durations using a given track as a "template" of sorts.

**Comment by Danielku15** [comment]
I am not sure if I understand your problem & goal from this very solution oriented proposal. If you explain a bit more on what you are trying to achieve a better discussion would be possible. `playNote` is intended to be used with notes from the currently rendered and "active" Score. 

It sounds like a very specific need from you and I am not sure if this should really be part of alphaTab. 

You might also just generate an alphaTex string on your side and then use the [`ScoreImporter`](https://alphatab.net/docs/guides/lowlevel-apis#loading-files-via-scoreloader) to create a score out of it and load the list of notes from it for further usage somehow.

**Reply by kyledecot** [reply]
I was wanting to create six different `<input type="button" value="{NOTE_NAME}" /> for each string of the currently rendered track. When clicked I was wanting to play a whole note to serve as a very primitive tuner. 

I had thought of using `ScoreImporter` as you had suggested however this would require me to "unload" the current score.

I agree that AlphaTab is probably not the right level for something like this to live at but could AlphaSynth be used to accomplish this?

Thanks for your reply!

**Reply by Danielku15** [reply]
I think something got lost in your message. You want to create six different what? 😅

Maybe you can describe me from an end-user experience what you want to build and then relate what technical level things you want to use to build it. 

What I undestood until now: You want to build a Guitar Tuner which allows the users to tune the guitar "by ear". So you want to build a guitar tuner UI, and when the string is pressed, you want to play some note with the right program and tune of the string. For this you want to use alphaTab/alphaSynth. 

This should be possible with the APIs of today. I could fiddle you an example if you confirm that this is what you want to build. 

AlphaSynth has internally an API called `playOneTimeMidiFile` which stops the current midi playback of the synth and plays the given file once. 
https://github.com/CoderLine/alphaTab/blob/43cbd1f78bf16a4d8a8d086592fd9f278a23d14a/src/synth/IAlphaSynth.ts#L113

This is used when asking for a playback of a single beat or note and you would do the same in your tuner code: 
https://github.com/CoderLine/alphaTab/blob/43cbd1f78bf16a4d8a8d086592fd9f278a23d14a/src/AlphaTabApiBase.ts#L698-L714

1. Setup a new alphaTab instance for the tuner. You could also use the instance you have for displaying in case you don't care that the tuner interferes with the main playback. 
2. Create a `MidiFile` for the playback. 
3. Send it to the synth for playback through `player.playOneTimeMidiFile`

Available options for this workflow are

a) Create a new full `Score` with a single note inside yourself and call `playNote` or `playBeat` with it. Technically it seems you can send in a beat/note from a different Score to an alphaTab instance and it will work correctly. 
b) Create an alphaTex string and use the `ScoreLoader` to parse it into a `Score` and send it to `playNote` or `playBeat`
c) Use the low level APIs to generate the midi yourself and send it to a player instance. Not everything might be fully exposed from the internal classes. I would have to check what needs to be opened for this workflow. 


I once tried to reduce the public API surface to be able to follow semantic versioning easier and have the possibility to change internal APIs. I might need to change my strategy to rather mark some APIs as internal and subject to break even in minor versions. But this way I could expose more for the "expert/advanced" users.
